kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/ksp/TargetTypes.kt

```diff
@@ -0,0 +1,280 @@
+    return null
+  }
+  logger.check(type.classKind != ClassKind.ENUM_CLASS, type) {
+    "@JsonClass with 'generateAdapter = \"true\"' can't be applied to ${type.qualifiedName?.asString()}: code gen for enums is not supported or necessary"
+  }
+  logger.check(type.classKind == CLASS && type.origin == Origin.KOTLIN, type) {
+    "@JsonClass can't be applied to ${type.qualifiedName?.asString()}: must be a Kotlin class"
+  }
+  logger.check(Modifier.INNER !in type.modifiers, type) {
+    "@JsonClass can't be applied to ${type.qualifiedName?.asString()}: must not be an inner class"
+  }
+  logger.check(Modifier.SEALED !in type.modifiers, type) {
+    "@JsonClass can't be applied to ${type.qualifiedName?.asString()}: must not be sealed"
+  }
+  logger.check(Modifier.ABSTRACT !in type.modifiers, type) {
+    "@JsonClass can't be applied to ${type.qualifiedName?.asString()}: must not be abstract"
+  }
+  logger.check(!type.isLocal(), type) {
+    "@JsonClass can't be applied to ${type.qualifiedName?.asString()}: must not be local"
+  }
+  logger.check(type.isPublic() || type.isInternal(), type) {
+    "@JsonClass can't be applied to ${type.qualifiedName?.asString()}: must be internal or public"
+  }
+
+  val classTypeParamsResolver = type.typeParameters.toTypeParameterResolver(
+    sourceTypeHint = type.qualifiedName!!.asString()
+  )
+  val typeVariables = type.typeParameters.map { it.toTypeVariableName(classTypeParamsResolver) }
+  val appliedType = AppliedType.get(type)
+
+  val constructor = primaryConstructor(resolver, type, classTypeParamsResolver, logger)
+    ?: run {
+      logger.error("No primary constructor found on $type", type)
+      return null
+    }
+  if (constructor.visibility != KModifier.INTERNAL && constructor.visibility != KModifier.PUBLIC) {
+    logger.error(
+      "@JsonClass can't be applied to $type: " +
+        "primary constructor is not internal or public",
+      type
+    )
+    return null
+  }
+
+  val properties = mutableMapOf<String, TargetProperty>()
+
+  val originalType = appliedType.type
+  for (supertype in appliedType.supertypes(resolver)) {
+    val classDecl = supertype.type
+    if (!classDecl.isKotlinClass()) {
+      logger.error(
+        """
+        @JsonClass can't be applied to $type: supertype $supertype is not a Kotlin type.
+        Origin=${classDecl.origin}
+        Annotations=${classDecl.annotations.joinToString(prefix = "[", postfix = "]") { it.shortName.getShortName() }}
+        """.trimIndent(),
+        type
+      )
+      return null
+    }
+    val supertypeProperties = declaredProperties(
+      constructor = constructor,
+      originalType = originalType,
+      classDecl = classDecl,
+      resolver = resolver,
+      typeParameterResolver = classDecl.typeParameters
+        .toTypeParameterResolver(classTypeParamsResolver)
+    )
+    for ((name, property) in supertypeProperties) {
+      properties.putIfAbsent(name, property)
```

Comment:
Could there be a collision among private property names? a private property name